Panic during directory scan

**What happened**: I ran `grype .` in my local "syft" repo directory, and grype exited with a panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x18ea1a1]

goroutine 67 [running]:
github.com/anchore/grype/grype/matcher/apk.(*Matcher).cpeMatchesWithoutSecDbFixes(0x2329a98, 0x1d8a4b0, 0xc013c646e0, 0x0, 0xc0239fb0e0, 0x24, 0xc023589c02, 0xa, 0xc023589c12, 0x8, ...)
        /Users/runner/work/grype/grype/grype/matcher/apk/matcher.go:60 +0x101
github.com/anchore/grype/grype/matcher/apk.(*Matcher).Match(0x2329a98, 0x1d8a4b0, 0xc013c646e0, 0x0, 0xc0239fb0e0, 0x24, 0xc023589c02, 0xa, 0xc023589c12, 0x8, ...)
        /Users/runner/work/grype/grype/grype/matcher/apk/matcher.go:35 +0x199
github.com/anchore/grype/grype/matcher.(*controller).findMatches(0x22f5668, 0x1d8a4b0, 0xc013c646e0, 0x0, 0xc025280000, 0x29d, 0x29d, 0xc000183400)
        /Users/runner/work/grype/grype/grype/matcher/controller.go:95 +0x3b5
github.com/anchore/grype/grype/matcher.FindMatches(...)
        /Users/runner/work/grype/grype/grype/matcher/controller.go:113
github.com/anchore/grype/grype.FindVulnerabilitiesForPackage(...)
        /Users/runner/work/grype/grype/grype/lib.go:28
github.com/anchore/grype/cmd.startWorker.func1(0xc0007d4540, 0x7ffeefbff6e6, 0x1, 0x0)
        /Users/runner/work/grype/grype/cmd/root.go:231 +0x385
created by github.com/anchore/grype/cmd.startWorker
        /Users/runner/work/grype/grype/cmd/root.go:170 +0x76
```

**What you expected to happen**: A successful execution that shows vulnerability matches for my "syft" directory.

**How to reproduce it (as minimally and precisely as possible)**: Still working on this. Thus far, I've only seen it for _directory_ scans, and not all directories trigger the panic.


**Environment**:
- Output of `grype version`:

```
Application:          grype
Version:              0.13.0
BuildDate:            2021-06-02T01:57:12Z
GitCommit:            3d21b8397d65770d292184b09a4f676bce6f3ec8
GitTreeState:         clean
Platform:             darwin/amd64
GoVersion:            go1.16.4
Compiler:             gc
Supported DB Schema:  3
```

- OS (e.g: `cat /etc/os-release` or similar): macOS 11.4

--- a/grype/matcher/apk/matcher.go
+++ b/grype/matcher/apk/matcher.go
@@ -57,7 +57,7 @@ func (m *Matcher) cpeMatchesWithoutSecDbFixes(store vulnerability.Provider, d *d
 
 	// remove cpe matches where there is an entry in the secDB for the particular package-vulnerability pairing, and the
 	// installed package version is >= the fixed in version for the secDB record.
-	secDbVulnerabilities, err := store.GetByDistro(*d, p)
+	secDbVulnerabilities, err := store.GetByDistro(d, p)
 	if err != nil {
 		return nil, err
 	}

--- a/grype/matcher/common/distro_matchers.go
+++ b/grype/matcher/common/distro_matchers.go
@@ -22,7 +22,7 @@ func FindMatchesByPackageDistro(store vulnerability.ProviderByDistro, d *distro.
 
 	var allPkgVulns []vulnerability.Vulnerability
 
-	allPkgVulns, err = store.GetByDistro(*d, p)
+	allPkgVulns, err = store.GetByDistro(d, p)
 	if err != nil {
 		return nil, fmt.Errorf("matcher failed to fetch distro='%s' pkg='%s': %w", d, p.Name, err)
 	}

--- a/grype/matcher/common/distro_matchers_test.go
+++ b/grype/matcher/common/distro_matchers_test.go
@@ -55,7 +55,7 @@ func (pr *mockDistroProvider) stub() {
 	}
 }
 
-func (pr *mockDistroProvider) GetByDistro(d distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
+func (pr *mockDistroProvider) GetByDistro(d *distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
 	return pr.data[strings.ToLower(d.Type.String())+":"+d.FullVersion()][p.Name], nil
 }
 

--- a/grype/matcher/dpkg/matcher_mocks_test.go
+++ b/grype/matcher/dpkg/matcher_mocks_test.go
@@ -50,6 +50,6 @@ func (pr *mockProvider) stub() {
 	}
 }
 
-func (pr *mockProvider) GetByDistro(d distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
+func (pr *mockProvider) GetByDistro(d *distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
 	return pr.data[strings.ToLower(d.Type.String())+":"+d.FullVersion()][p.Name], nil
 }

--- a/grype/matcher/rpmdb/matcher_mocks_test.go
+++ b/grype/matcher/rpmdb/matcher_mocks_test.go
@@ -50,7 +50,7 @@ func (pr *mockProvider) stub() {
 	}
 }
 
-func (pr *mockProvider) GetByDistro(d distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
+func (pr *mockProvider) GetByDistro(d *distro.Distro, p pkg.Package) ([]vulnerability.Vulnerability, error) {
 	var ty = strings.ToLower(d.Type.String())
 	if d.Type == distro.CentOS || d.Type == distro.RedHat {
 		ty = "rhel"

--- a/grype/vulnerability/provider.go
+++ b/grype/vulnerability/provider.go
@@ -13,7 +13,7 @@ type Provider interface {
 }
 
 type ProviderByDistro interface {
-	GetByDistro(distro.Distro, pkg.Package) ([]Vulnerability, error)
+	GetByDistro(*distro.Distro, pkg.Package) ([]Vulnerability, error)
 }
 
 type ProviderByLanguage interface {

--- a/grype/vulnerability/store_adapter.go
+++ b/grype/vulnerability/store_adapter.go
@@ -21,14 +21,18 @@ func NewProviderFromStore(store db.VulnerabilityStoreReader) *StoreAdapter {
 	}
 }
 
-func (pr *StoreAdapter) GetByDistro(d distro.Distro, p pkg.Package) ([]Vulnerability, error) {
-	vulns := make([]Vulnerability, 0)
+func (pr *StoreAdapter) GetByDistro(d *distro.Distro, p pkg.Package) ([]Vulnerability, error) {
+	if d == nil {
+		return nil, nil
+	}
 
-	namespace := distroNamespace(d)
+	namespace := distroNamespace(*d)
 	allPkgVulns, err := pr.store.GetVulnerability(namespace, p.Name)
 	if err != nil {
 		return nil, fmt.Errorf("provider failed to fetch namespace='%s' pkg='%s': %w", namespace, p.Name, err)
 	}
+
+	var vulnerabilities []Vulnerability
 
 	for _, vuln := range allPkgVulns {
 		vulnObj, err := NewVulnerability(vuln)
@@ -36,10 +40,10 @@ func (pr *StoreAdapter) GetByDistro(d distro.Distro, p pkg.Package) ([]Vulnerabi
 			return nil, fmt.Errorf("provider failed to parse distro='%s': %w", d, err)
 		}
 
-		vulns = append(vulns, *vulnObj)
+		vulnerabilities = append(vulnerabilities, *vulnObj)
 	}
 
-	return vulns, nil
+	return vulnerabilities, nil
 }
 
 func (pr *StoreAdapter) GetByLanguage(l syftPkg.Language, p pkg.Package) ([]Vulnerability, error) {

--- a/grype/vulnerability/store_adapter_test.go
+++ b/grype/vulnerability/store_adapter_test.go
@@ -54,7 +54,19 @@ func TestGetByDistro(t *testing.T) {
 			t.Errorf("diff: %+v", d)
 		}
 	}
+}
 
+func TestGetByDistro_nilDistro(t *testing.T) {
+	provider := NewProviderFromStore(newMockStore())
+
+	p := pkg.Package{
+		Name: "neutron",
+	}
+
+	vulnerabilities, err := provider.GetByDistro(nil, p)
+
+	assert.Empty(t, vulnerabilities)
+	assert.NoError(t, err)
 }
 
 func must(c syftPkg.CPE, e error) syftPkg.CPE {

--- a/grype/vulnerability/store_adapter_test.go
+++ b/grype/vulnerability/store_adapter_test.go
@@ -25,7 +25,7 @@ func TestGetByDistro(t *testing.T) {
 		Name: "neutron",
 	}
 
-	actual, err := provider.GetByDistro(d, p)
+	actual, err := provider.GetByDistro(&d, p)
 	if err != nil {
 		t.Fatalf("failed to get by distro: %+v", err)
 	}

